pkg/networkservice/common/connect/server.go

```diff
@@ -21,247 +19,112 @@ package connect
 	}
 	if clientErr != nil {
 		return rv, errors.Wrap(clientErr, "errors during client close")
 	}
-
 	return rv, err
 }
 
-func (c *connectServer) findOrCreateClient(ctx context.Context, clientURI *url.URL, conn *networkservice.Connection) (ce *clientEntry, err error) {
-	connectionExists := false
-	<-c.executor.AsyncExec(func() {
-		ce, connectionExists = c.clients[clientURI.String()]
-		if !connectionExists {
-			ce = &clientEntry{
-				clientURI:   clientURI,
-				ready:       make(chan struct{}),
-				connections: map[string]*networkservice.Connection{},
-			}
-			// Put/Update connection
-			c.clients[clientURI.String()] = ce
+func (c *connectServer) client(ctx context.Context, conn *networkservice.Connection) networkservice.NetworkServiceClient {
+	// Fast path, if we have a client for this conn.GetId(), use it
+	client, _ := c.clientsByID.Load(conn.GetId())
+
+	// If we didn't find a client, we fall back to clientURL
+	if client == nil {
+		clientURL := clienturl.ClientURL(ctx)
+		// If we don't have a clientURL, all we can do is return errors
+		if clientURL == nil {
+			clientErr := errors.Errorf("clientURL not found for incoming connection: %+v", conn)
+			return injecterror.NewClient(clientErr)
 		}
-		ce.connections[conn.Id] = conn
-	})
-
-	if connectionExists {
-		// Wait until connection is ready, or context timeout
-		select {
-		case <-ctx.Done():
-			// We failed, we need to remove pending state and close connection if we are last one.
-			<-c.executor.AsyncExec(func() {
-				delete(ce.connections, conn.Id)
-				c.closeClient(ctx, ce)
-			})
-			return nil, errors.Errorf("context timeout")
-		case <-ce.ready:
-			// all is fine, just return
-			if ce.err != nil {
-				<-c.executor.AsyncExec(func() {
-					delete(ce.connections, conn.Id)
-					c.closeClient(ctx, ce)
-				})
-				return nil, ce.err
+		// Fast path: load the client by URL.  In the unfortunate event someone has poorly chosen
+		// a clientFactory or dialOptions that take time, this will be faster than
+		// creating a new one and doing a LoadOrStore
+		client, _ = c.clientsByURL.Load(clientURL.String())
+		// If we still don't have a client, create one, and LoadOrStore it
+		if client == nil {
+			// Note: clienturl.NewClient(...) will get properly cleaned up when dereferences
+			client = clienturl.NewClient(clienturl.WithClientURL(c.ctx, clientURL), c.clientFactory, c.clientDialOptions...)
+			client, _ = c.clientsByURL.LoadOrStore(clientURL.String(), client)
+			// Wrap the client in a per-connection connect.NewClient(...)
+			// when this client receive a 'Close' it will call the cancelFunc provided deleting it from the various
+			// maps.
+			client = chain.NewNetworkServiceClient(
+				NewClient(func() {
+					c.clientsByID.Delete(conn.GetId())
+					c.clientsByURL.Delete(clientURL.String())
+				}),
+				client,
+			)
+			var loaded bool
+			client, loaded = c.clientsByID.LoadOrStore(conn.GetId(), client)
```

Comment:
Even if we get near simultaneous initial requests for the same Connection.Id causing two separate cases where client == nil in line 93... they would both hit this point, where precisely *one* of them would win on the LoadOrStore.  One would be Stored, the other would be Loaded.  If you see the comment below about why we have the if loaded {...} its actually cleaning up after the clientByURL refcount for that case.

So even in the *extremely* unlikely case of nearly simultaneous initial Requests... we maintain correctness, at a price much much smaller than global mutexes.  And in the common case without this collision, enormously smaller than global mutexes.